### ISSUE TITLE
Design: Engine OpenGL extension capability (DEF-2886)

### ISSUE THREAD

**Issue by britzl** [defold/defold#3350]
To stay current, and enable our users to use newer rendering techniques than is possible for OpenGL ES2, we need to enable the user to take advantage of OpenGL extensions.

Requirement involves Api-features to check if an extension is available and to be able to enable it (shaders).

This also involves being able to dynamically select shaders depending on extension availability etc. We can not rely on text-based shaders as we might want to implement other rendering api's only allowing for binary run-time shaders.

Initially we wanted to investigate the possibility of Vulkan and OpenGL ES3 implementations, but this is nothing we will do at this time.
 * Vulcan is not supported on all hardware we support (far from it). It would only serve as an optimised extension at best, so it's not going to happen in short time perspective. Furthermore, costs for using MoltenVK is yet unknown (Apple does not support Vulcan directly)
 * The same goes for OpenGL ES3, although its supported by more hardware and can _potentially_ be implemented in parallel with the ES2 codepath. We will currently not invest in such an implementation though.
2017-09-15 10:52:15 (sven.andersson)
Assigning to [~andreas.tadic] for more input and discussion in the document.

**Comment by AGulev** [comment]
closed as duplicate of https://github.com/defold/defold/issues/3790
